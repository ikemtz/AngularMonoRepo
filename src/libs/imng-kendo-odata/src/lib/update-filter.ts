import { CompositeFilterDescriptor, FilterDescriptor, isCompositeFilterDescriptor } from '@progress/kendo-data-query';

export function updateFilter(compositefilter: CompositeFilterDescriptor, newFilter: FilterDescriptor): CompositeFilterDescriptor {
  compositefilter.filters.forEach(subFilter => {
    if (isCompositeFilterDescriptor(subFilter)) {
      updateFilter(subFilter, newFilter);
    } else if (subFilter.field === newFilter.field) {
      compositefilter = {
        ...compositefilter,
        filters: [...compositefilter.filters.filter(f => isCompositeFilterDescriptor(f) ||
          f.field !== newFilter.field), newFilter]
      };
    }
  });
  return compositefilter;
}
