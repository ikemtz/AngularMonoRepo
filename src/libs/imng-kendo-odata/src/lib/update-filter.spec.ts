import { CompositeFilterDescriptor, FilterDescriptor } from '@progress/kendo-data-query';
import { updateFilter } from './update-filter';

describe('updateFilter', () => {
  const genFilter = (x: string): FilterDescriptor => ({ field: x, operator: 'eq', value: 2 });
  it('should update subFilters', () => {
    const filters: CompositeFilterDescriptor = {
      logic: 'and',
      filters: [genFilter('y'), genFilter('z'),
      { filters: [genFilter('a'), genFilter('b')], logic: 'and' }],
    };
    const result = updateFilter(filters, { ...genFilter('b'), value: 'xyz' });
    expect(result).toMatchSnapshot();
  });
  it('should update filters', () => {
    const filters: CompositeFilterDescriptor = {
      logic: 'and',
      filters: [genFilter('y'), genFilter('z'), genFilter('b')],
    };
    const result = updateFilter(filters, { ...genFilter('b'), value: 'xyz' });
    expect(result).toMatchSnapshot();
  });
});
