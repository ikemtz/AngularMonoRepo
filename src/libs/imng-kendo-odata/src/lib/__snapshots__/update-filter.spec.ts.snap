// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateFilter should update filters 1`] = `
{
  "filters": [
    {
      "field": "y",
      "operator": "eq",
      "value": 2,
    },
    {
      "field": "z",
      "operator": "eq",
      "value": 2,
    },
    {
      "field": "b",
      "operator": "eq",
      "value": "xyz",
    },
  ],
  "logic": "and",
}
`;

exports[`updateFilter should update subFilters 1`] = `
{
  "filters": [
    {
      "field": "y",
      "operator": "eq",
      "value": 2,
    },
    {
      "field": "z",
      "operator": "eq",
      "value": 2,
    },
    {
      "filters": [
        {
          "field": "a",
          "operator": "eq",
          "value": 2,
        },
        {
          "field": "b",
          "operator": "eq",
          "value": 2,
        },
      ],
      "logic": "and",
    },
  ],
  "logic": "and",
}
`;
